### ISSUE TITLE
require-atomic-updates false positive with inline functions

### ISSUE THREAD

**Issue by Standard8** [eslint/eslint#11723]
**Tell us about your environment**

* **ESLint Version:** 6.0.0-alpha.1
* **Node Version:** 8.16.0
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    "parserOptions": {
        "ecmaVersion": 2017
    },
    "rules": {
        "require-atomic-updates": "error",
    },
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
function get() { return {}; }
function update() {}

async function foo(bar) {
  let oldBar = await get(bar.id);

  let doIt = () => {
    bar.visible = true;
  };

  if (oldBar) {
    bar.installDate = oldBar.installDate;
  } else {
    bar.installDate = new Date();
  }
}
```

```bash
./node_modules/.bin/eslint index.js
```

**What did you expect to happen?**

No errors

**What actually happened? Please include the actual, raw output from ESLint.**

```
index.js
  12:5  error  Possible race condition: `bar.installDate` might be reassigned based on an outdated value of `bar.installDate`  require-atomic-updates
  15:5  error  Possible race condition: `bar.installDate` might be reassigned based on an outdated value of `bar.installDate`  require-atomic-updates

✖ 2 problems (2 errors, 0 warnings)
```

The problem in this case appears to be the inline function. If I comment out the `bar.visible = true;` then the errors go away.

**Are you willing to submit a pull request to fix this bug?**
Not at the moment

**Comment by mysticatea** [comment]
Thank you for your report.

I confirmed it.

Hmm. tough to know what's happened 😅 

The current `require-atomic-updates` looks aggressive. If one or more closures used a variable, the rule addresses the variable to be a danger. I'm considering how the rule should address closures.

**Comment by mysticatea** [comment]
I'm leaning towards that we should report the `bar.installDate` consistently because the `bar` is a parameter. If we called the function `foo` twice in parallel, it would cause a race condition.

**Comment by lll000111** [comment]
I'm confused by the report and the fix (PR) and the comments there.

In this issues based on function parameters I found this inconsistency &mdash; no matter what one thinks of the validity of the error report itself, it is inconsistent between the examples:

This error comes up in demo function 1:

```
8:5  error  Possible race condition: `planObj.endTime` might be reassigned based on an outdated value of `planObj.endTime`  require-atomic-updates
```

Code:

```js
'use strict';

import {someSyncFunction, wait} from 'somewhere';

async function doStuff1(planObj) {
    someSyncFunction(planObj);
    await wait(100);
    planObj.endTime = Date.now();  // 8:5: ESLINT ERROR
}

async function doStuff2(planObj) {
    await wait(100);
    planObj.endTime = Date.now();
}

async function doStuff3(planObj) {
    const planObjPointerCopy = planObj;
    someSyncFunction(planObj);
    await wait(100);
    planObjPointerCopy.endTime = Date.now();
}
```

The third example is no better than the first one, yet there is no error.

Of course, I think there is no reason for there to be an error even in the first function. Even if the `someSyncFunction` were to mutate the object, what of it? The following code is not reading but writing!

**Comment by lll000111** [comment]
In any case, the opinions about what the rule is supposed to catch made here are much(!) more than what the documentation says. The example there is simple and straightforward, what @mysticatea says above is much broader.

So I would like to add that the documentation should be updated too.
